pkg/scalers/datadog_scaler.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	return apiClient, nil
+}
+
+// No need to close connections
+func (s *datadogScaler) Close(context.Context) error {
+	return nil
+}
+
+// IsActive returns true if we are able to get metrics from Datadog
+func (s *datadogScaler) IsActive(ctx context.Context) (bool, error) {
+	ctx = context.WithValue(
+		ctx,
+		datadog.ContextAPIKeys,
+		map[string]datadog.APIKey{
+			"apiKeyAuth": {
+				Key: s.metadata.apiKey,
+			},
+			"appKeyAuth": {
+				Key: s.metadata.appKey,
+			},
+		},
+	)
+
+	ctx = context.WithValue(ctx,
+		datadog.ContextServerVariables,
+		map[string]string{
+			"site": s.metadata.datadogSite,
+		})
+
+	resp, _, err := s.apiClient.MetricsApi.QueryMetrics(ctx, time.Now().Unix()-int64(s.metadata.age), time.Now().Unix(), s.metadata.query)
+
+	if err != nil {
+		return false, err
+	}
+
+	series := resp.GetSeries()
+
+	if len(series) == 0 {
+		return false, nil
+	}
+
+	points := series[0].GetPointlist()
+
+	if len(points) == 0 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// getQueryResult returns result of the scaler query
+func (s *datadogScaler) getQueryResult(ctx context.Context) (int, error) {
+	ctx = context.WithValue(
+		ctx,
+		datadog.ContextAPIKeys,
+		map[string]datadog.APIKey{
+			"apiKeyAuth": {
+				Key: s.metadata.apiKey,
+			},
+			"appKeyAuth": {
+				Key: s.metadata.appKey,
+			},
+		},
+	)
+
+	ctx = context.WithValue(ctx,
+		datadog.ContextServerVariables,
+		map[string]string{
+			"site": s.metadata.datadogSite,
+		})
+
+	resp, _, err := s.apiClient.MetricsApi.QueryMetrics(ctx, time.Now().Unix()-int64(s.metadata.age), time.Now().Unix(), s.metadata.query)
+	if err != nil {
+		return -1, fmt.Errorf("error when retrieving Datadog metrics: %s", err)
+	}
+
+	series := resp.GetSeries()
+
+	if len(series) == 0 {
+		return 0, fmt.Errorf("no Datadog metrics returned")
+	}
+
+	points := series[0].GetPointlist()
+
+	if len(points) == 0 || len(points[0]) < 2 {
+		return 0, fmt.Errorf("no Datadog metrics returned")
+	}
+
+	return int(*points[0][1]), nil
+}
+
+// GetMetricSpecForScaling returns the MetricSpec for the Horizontal Pod Autoscaler
+func (s *datadogScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	var externalMetric *v2beta2.ExternalMetricSource
+	externalMetric = new(v2beta2.ExternalMetricSource)
+
+	targetQueryValue := resource.NewQuantity(int64(s.metadata.queryValue), resource.DecimalSI)
+
+	switch s.metadata.vType {
+	case average:
+		externalMetric = &v2beta2.ExternalMetricSource{
+			Metric: v2beta2.MetricIdentifier{
+				Name: s.metadata.metricName,
+			},
+			Target: v2beta2.MetricTarget{
+				Type:         v2beta2.AverageValueMetricType,
+				AverageValue: targetQueryValue,
+			},
+		}
+	case global:
+		externalMetric = &v2beta2.ExternalMetricSource{
+			Metric: v2beta2.MetricIdentifier{
+				Name: s.metadata.metricName,
+			},
+			Target: v2beta2.MetricTarget{
+				Type:  v2beta2.ValueMetricType,
+				Value: targetQueryValue,
+			},
+		}
```

Comment:
do we need to fail here if there is no value value for `s.metadata.vType`? If, for some reason, it's not `average` or `global`, `externalMetric` will be `nil`

Reply:
the type is defaulted to Average, so I it won't be empty. In any case, you mentioned this in your previous review, and that's why I added the "new" constructor, as in that case, the pointer won't be nil: https://go.dev/play/p/ROR4tfKKQl0